feature to report installed apps stats on mac, windows, and linux

rohan-coder02 opened the pull request:
### **Title:**

Add Installed Apps API for Windows, Linux, and macOS (#2581)

### **Description:**

This PR introduces a new Installed Applications API for fetching application details across Windows, Linux, and macOS platforms. 

### **Changes Made:**

- Introduced a common ApplicationInfo interface with the following fields:
- name, version, vendor, and installDate (epoch time).
- A key-value map additionalInfo to handle platform-specific details.

**_Windows Implementation:_**

Fetching app details from Windows Registry 

**_Linux Implementation:_**

Support for both dpkg (Debian-based) and rpm (Red Hat-based) package managers.

**_macOS Implementation:_**

Parsing data from system_profiler.

### Testing Done:

Windows, MacOS, Linux (Redhat, Ubuntu, Amazon Linux 2, CentOS)

dbwiddis commented on oshi-core/src/main/java/oshi/util/ParseUtil.java:
```diff
@@ -1352,4 +1356,57 @@ public static String getValueOrUnknown(Map<String, String> map, String key) {
+    public static long parseWindowsDateToEpoch(String dateString) {
+        try {
+            DateTimeFormatter windowsFormatter = DateTimeFormatter.ofPattern("yyyyMMdd", Locale.US);
+            LocalDateTime localDateTime = LocalDate.parse(dateString, windowsFormatter).atStartOfDay();
+            Instant instant = localDateTime.atZone(ZoneId.systemDefault()).toInstant();
+            return instant.toEpochMilli();
```
you're repeating these lines in both new methods.  They should be combined (I'm surprised they don't already exist somewhere in this util).

Looks like we parse from `LocalDateTime` to millis in the `who` util, so maybe these can be combined and new format(s) added.
https://github.com/oshi/oshi/blob/8be455875c900e409e7caeaf4fc3d8d18df81f4c/oshi-core/src/main/java/oshi/driver/unix/Who.java#L106-L112

rohan-coder02 replied:
Thanks for the feedback.

I've created a common function -> parseDateToEpoch, in ParseUtil class that handles date from Windows (yyyyMMdd) and Mac (dd/MM/yy, HH:mm)

dbwiddis reviewed (commented):
Great start!  I think there's room to simplify a lot of this.

rohan-coder02 commented on oshi-core/src/main/java/oshi/driver/windows/registry/InstalledAppsData.java:
```diff
@@ -0,0 +1,91 @@
+
+        for (String path : UNINSTALL_PATHS) {
+            List<String> output = ExecutingCommand.runNative("reg query " + path + " /s");
+            parseRegistryOutput(output, appInfoList);
```
Thanks for the feedback.

I've used JNA Advapi32Util functions, added the required registry paths and was able to installed softwares as seen previously.

rohan-coder02 commented on oshi-core/src/main/java/oshi/software/os/AppInfo.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.util.Map;
+
+public interface AppInfo {
```
I've accordingly modified

rohan-coder02 commented on oshi-core/src/main/java/oshi/software/os/AppInfo.java:
```diff
@@ -0,0 +1,19 @@
+
+    String getVendor();
+
+    long getLastModifiedEpoch();
```
I've accordingly modified

rohan-coder02 commented on oshi-core/src/main/java/oshi/software/os/InstalledAppsFactory.java:
```diff
@@ -0,0 +1,26 @@
+public class InstalledAppsFactory {
+
+    public static AbstractInstalledApps getInstalledApps() {
+        String os = System.getProperty("os.name").toLowerCase();
```
Thanks for the feedback.

I've accordingly modified.

Java apps using OSHI can call this API as below

```
SystemInfo si = new SystemInfo();
List<ApplicationInfo> apps = si.getOperatingSystem().getInstalledApplications();
```

rohan-coder02 commented on oshi-core/src/main/java/oshi/util/ParseUtil.java:
```diff
@@ -1352,4 +1356,57 @@ public static String getValueOrUnknown(Map<String, String> map, String key) {
+     */
+    public static long parseMacDateToEpoch(String dateString) {
+        try {
+            DateTimeFormatter macFormatter = DateTimeFormatter.ofPattern("dd/MM/yy, HH:mm", Locale.US);
```
I've updated the code as Locale.ROOT

rohan-coder02 commented on oshi-core/src/main/java/oshi/software/os/linux/LinuxInstalledApps.java:
```diff
@@ -0,0 +1,93 @@
+            command = "rpm -qa --queryformat '%{NAME}|%{VERSION}-%{RELEASE}|%{ARCH}|%{SIZE}|%{INSTALLTIME}|%{PACKAGER}|%{SOURCERPM}|%{URL}\\n'";
+            break;
+        default:
+            LOG.warn("Installed Apps stats are not supported for this flavor of Linux OS");
```
I've included the doc

dbwiddis reviewed (commented):
Thanks for iterating on this.  It's looking really good.

I have been a bit busy today and have mostly glanced at it, just a general comment that there still seems to be a lot more classes than needed to populate the ApplicationInfo objects.   

I'm thinking that we may just need a concrete ApplicationInfo class and not an interface, and then each OS could populate it with its own methods.  We don't need getters and toStrings on application-specific objects.

One other change I'd suggest is to use a memoizer for fetching this data.  It's not likely to change that often, so all the fetching code should be under a "query" method while the "get" method fetches the value from the memoizer.

Here's user group info with a 5-minute refresh, as an example.  Maybe 1 minute is appropriate here?  30 seconds?  probably longer than 1 second though.  https://github.com/oshi/oshi/blob/8be455875c900e409e7caeaf4fc3d8d18df81f4c/oshi-core/src/main/java/oshi/util/UserGroupInfo.java#L28

mprins commented:
I don't see this cater for alternate package managers that install software, eg. chocolatey, scoop, snap (Ubuntu's primary distribution channel), flatpack and/or docker? having incomplete data seems a set-up for disappointment

dbwiddis commented:
> I don't see this cater for alternate package managers 

I'm hoping the initial contribution with a useful API will attract more submissions from others who want their favorite package managers included and are best suited to make those contributions.  MacOS needs Homebrew and MacPorts as well.  All in good time. 

I don't think we need to try to include everything in https://en.wikipedia.org/wiki/List_of_software_package_management_systems in the first PR.

> having incomplete data seems a set-up for disappointment

Having no data is also disappointing.  This feature has been requested by multiple people.  And additional options can be added.

rohan-coder02 commented:
Hi @dbwiddis 

Thanks for the feedback.

The recent commit does the follows:

- Removing redundant OS-specific classes & making ApplicationInfo a concrete class
- Memoizing installed apps query  (1-minute)

I had a question: Currently, the ApplicationInfo class includes getter methods and a toString() function. Should these be removed?

dbwiddis commented on oshi-core/src/main/java/oshi/software/os/windows/WindowsInstalledApps.java:
```diff
@@ -0,0 +1,44 @@
+        List<Map<InstalledAppsData.AppInfo, String>> registryResults = InstalledAppsData.queryInstalledApps();
+
+        for (Map<InstalledAppsData.AppInfo, String> app : registryResults) {
+            String installDate = ParseUtil.getValueOrUnknown(app, InstalledAppsData.AppInfo.INSTALLDATE);
+            long installDateEpoch = 0;
+            if (!installDate.equals(Constants.UNKNOWN)) {
+                installDateEpoch = ParseUtil.parseDateToEpoch(installDate, "yyyyMMdd");
+            }
```
This seems a strange pattern to create a new method to use UNKNOWN as a default and then check it, rather than just use `app.getOrDefault()` with an empty string (or even null) default that you can then check for with `Util.isBlank()` (or even `isBlankOrUnknown()`!).  Or just send it to the parse method and do some sanity checking of the format before parsing.  And offer a default from that method rather than defaulting to 0.

rohan-coder02 commented on oshi-core/src/main/java/oshi/driver/windows/registry/InstalledAppsData.java:
```diff
@@ -0,0 +1,71 @@
+
+    public enum AppInfo {
+        NAME, VERSION, VENDOR, INSTALLDATE, INSTALLLOCATION, INSTALLSOURCE;
+    }
```
Thanks for the feedback.

In the recent commit, Refactored InstalledAppsData to fetch registry data and convert to ApplicationInfo

rohan-coder02 commented on oshi-core/src/main/java/oshi/software/os/ApplicationInfo.java:
```diff
@@ -0,0 +1,88 @@
+    private final long lastModifiedTime;
+    private final Map<String, String> additionalInfo;
+
+    public ApplicationInfo(String name, String version, String vendor, long lastModifiedTime, Map<String, String> additionalInfo) {
```
I've updated it to timestamp in ApplicationInfo class in the recent commit

rohan-coder02 commented on oshi-core/src/main/java/oshi/software/os/linux/LinuxInstalledApps.java:
```diff
@@ -0,0 +1,86 @@
+            command = "dpkg-query -W -f='${Package}|${Version}|${Architecture}|${Installed-Size}|${db-fsys:Last-Modified}|${Maintainer}|${Source}|${Homepage}\\n'";
+        } else if (isPackageManagerAvailable("rpm")) {
+            command = "rpm -qa --queryformat '%{NAME}|%{VERSION}-%{RELEASE}|%{ARCH}|%{SIZE}|%{INSTALLTIME}|%{PACKAGER}|%{SOURCERPM}|%{URL}\\n'";
+        } else {
```
Thanks for the feedback

I've updated it in the recent commit to include a map along with initializePackageManagerCommands function

rohan-coder02 commented on oshi-core/src/main/java/oshi/software/os/linux/LinuxOperatingSystem.java:
```diff
@@ -75,6 +79,9 @@ public class LinuxOperatingSystem extends AbstractOperatingSystem {
     public static final boolean HAS_SYSCALL_GETTID;
 
+    private final Supplier<List<ApplicationInfo>> installedAppsSupplier =
+        Memoizer.memoize(LinuxInstalledApps::queryInstalledApps, defaultExpiration());
```
Apologies, in the previous commit I misread the defaultExpiration to be 1 minute.

In the recent commit, I've added INSTALLED_APPS_EXPIRATION_NANOS in the Memoizer class whose time is 1 minute which is used by the suppliers in the respective OS specific classes

dbwiddis reviewed (commented):
Thanks for being patient with the feedback!

This is almost ready, just a few optimizations and we can get this merged.

dbwiddis commented:
Also some checkstyle errors:
> [INFO] There are 4 errors reported by Checkstyle 10.21.4 with config/checkstyle.xml ruleset.
> Error:  src/main/java/oshi/util/ParseUtil.java:[121,33] (naming) ConstantName: Name 'slash' must match pattern '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'.

There are already exclusions in the ruleset, this exclusion probably needs to have the line numbers adjusted for the 3 new import lines.

> Error:  src/main/java/oshi/software/os/linux/LinuxInstalledApps.java:[17,1] (design) HideUtilityClassConstructor: Utility classes should not have a public or default constructor.
> Error:  src/main/java/oshi/software/os/mac/MacInstalledApps.java:[17,1] (design) HideUtilityClassConstructor: Utility classes should not have a public or default constructor.
> Error:  src/main/java/oshi/software/os/windows/WindowsInstalledApps.java:[12,1] (design) HideUtilityClassConstructor: Utility classes should not have a public or default constructor.

These all need a private no-arg constructor.

rohan-coder02 commented on oshi-core/src/main/java/oshi/software/os/linux/LinuxInstalledApps.java:
```diff
@@ -0,0 +1,98 @@
+
+        for (String line : output) {
+            // split by the pipe character
+            String[] parts = line.split("\\|", -1); // -1 to keep empty fields
```
Thanks for the feedback.

I've updated it in the latest commit

rohan-coder02 commented on oshi-core/src/main/java/oshi/util/ParseUtil.java:
```diff
@@ -1352,4 +1355,60 @@ public static String getValueOrUnknown(Map<String, String> map, String key) {
+     */
+    public static String getValueOrUnknown(Map<?, String> map, Object key) {
+        String value = map.getOrDefault(key, "");
+        return value.isEmpty() ? Constants.UNKNOWN : value;
```
Thanks for the feedback.

I've updated it in the latest commit

rohan-coder02 commented on oshi-core/src/main/java/oshi/util/Memoizer.java:
```diff
@@ -19,13 +19,20 @@ public final class Memoizer {
     }
 
+    public static long installedAppsExpiration() {
+        return INSTALLED_APPS_EXPIRATION_NANOS.get();
```
Thanks for the feedback.

I've updated it in the latest commit

rohan-coder02 commented:
Added the log entry.

Regarding checkstyle errors

1)

> Error: src/main/java/oshi/util/ParseUtil.java:[121,33] (naming) ConstantName: Name 'slash' must match pattern '^[A-Z][A-Z0-9](_[A-Z0-9]+)$'.

This looks to be a part of below PR.
https://github.com/oshi/oshi/commit/80e49d4ed0e4448234722b7e17f75f8ed38ef0c2#diff-ae0f9f11563340129b17437e87131f5548220bfca5d9eaf437a577c51ed85841
Anyhow, I've modified `slash` -> `SLASH` to pass the checkstyle

2)

Yes, in the recent commit, I've made the constructor private in the installed apps class of Mac, Windows, and Linux

dbwiddis commented:
> Anyhow, I've modified `slash` -> `SLASH` to pass the checkstyle

As I me times this was an intentional style break. There is an exclusion for it but the three new import statements in this PR changed those lines.

rohan-coder02 commented:
> As I mentioned this was an intentional style break. There is an exclusion for it but the three new import statements in this PR changed those lines.

Understood.

I'll update this PR with `slash`.
